testbed/tests/e2e_test.go

```diff
@@ -0,0 +1,55 @@
+// Copyright 2019, OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package tests contains test cases. To run the tests go to tests directory and run:
+// TESTBED_CONFIG=local.yaml go test -v
+
+package tests
+
+import (
+	"fmt"
+	"testing"
+	"strconv"
+
+	"github.com/stretchr/testify/assert"
+
+	"github.com/open-telemetry/opentelemetry-service/testbed/testbed"
+)
+
+func TestBallastMemory(t *testing.T) {
+	tests := []struct{
+		ballastSize uint32
+		maxRSS uint32
+	}{
+		{100, 50},
+		{500, 70},
+		{1000, 100},
+	}
+
+	for _, test := range tests {
+		tc := testbed.NewTestCase(t, testbed.WithSkipResults())
+		tc.SetExpectedMaxRAM(test.maxRSS)
+
+		tc.StartAgent("--mem-ballast-size-mib", strconv.Itoa(int(test.ballastSize)))
+
+		rss, vms, err := tc.AgentMemoryInfo()
```

Comment:
You probably need to wait before querying memory usage, otherwise you may not get the expected results.

I would refactor this to (warning: pseudocode below)
```
WaitFor(AgentMemoryInfo().vms > test.ballastSize)
assert.True(rss <= test.maxRSS)
```
This would ensure you are measuring after ballast is acquired.

Reply:
Doesn't `StartAgent()` block until the agent child process starts up? I only need the process to start, it doesn't need to be ready to receive spans or anything like that because ballast is set at the very beginning and only thing we are testing here is if ballast is taking up RSS or only VMS.

Reply:
It waits for process to start, but doesn't wait for process to run and do anything that it is supposed to do, particularly reserve the memory ballast. You are basically racing here against ballast acquisition.

Reply:
OK. I think it's pretty hard or nearly impossible to hit this case but agree that it's better to add for the sake of correctness. Will do. Thanks.

Reply:
Added WaitFor call but 10 seconds was too long for this case IMO so I added another function that allows to specify a custom duration.

Reply:
WaitFor waits up to 10 seconds. Normally it should take 0 time, assuming the condition is true when WaitFor is called. It then probes the condition with exponentially increasing intervals.
Why is it not good enough?

Reply:
WaitFor will only waits full 10 seconds if the test is going to fail. I think that is fine, that should be very rare anyway, only when there is a bug.

If you absolutely want to control WaitFor maximum waiting duration I advise instead of introducing a new WaitForN function, make maximum waiting duration a TestCase setting. This keeps tests look cleaner and easier to understand.

But IMHO, it is overkill, just call the default WaitFor and be done with it.

Reply:
Normally it will take zero time but when the ballast code/test fails then the test will wait for 10 seconds which will be annoying when iterating over related code as in reality it'd either fail immediately or pass immediately. In no case would it not pass after ~1 second but work after 9. So just want to cut short the time and not wait for 10 seconds * number of cases when the test fails.

Reply:
OK, up to you. Fine either way with me.

Reply:
> If you absolutely want to control WaitFor maximum waiting duration I advise instead of introducing a new WaitForN function, make maximum waiting duration a TestCase setting. This keeps tests look cleaner and easier to understand.

To clarify the new function is just a one-liner wrapper over the existing one (plus swapped names). I think it's worth having a one line wrapper with default value over having to specify the default in all existing all sites but I'm fine with updating all call sites as well.

Reply:
SGTM.